### ISSUE TITLE
Data scaling and network stability

### ISSUE THREAD

**Issue by swigicat** [tiny-dnn/tiny-dnn#511]
Hi all,

I have a question regarding data scaling - my dataset is almost the same as CIFAR-10, only grayscale: 32x32, 1 channel, 4 classes, 3000 training and 800 test images. I'm using the CIFAR-10 example NN, with the number of channels set to 1, the rest is the same, but getting poor results - about 65% accuracy, while with tensorflow I get about 95%.

One reason I can think of is the data scaling. I have 12-bit data that is mostly background (i.e. 0), so if I scale that to -1..1, I get a large portion of the pixels sitting at -1. Is this a problem for your optimizer? With tensorflow I just feed the raw 12-bit data and it runs OK with it.

Another problem I'm facing is instability. When I get over 65% accuracy by changing things in the network structure, suddenly the accuracy drops to 25% and stays there for the remainder of the training. I tried various settings for learning rate, going down to 0.0001 and more epochs, it's more stable but then I get poor accuracy.

What do you think is the problem, should I change some other settings? Also, why the big discrepancy between this and tensorflow results with more or less the same NN structure, are they using fancier algorithms?

**Comment by Randl** [comment]
Have you got 95% with Tensorflow and 65% with tiny-dnn with same network architecture?

**Comment by swigicat** [comment]
Hi Randl,

No, the 65% comes from using the default tiny-dnn CIFAR-10 example, which I thought to be optimal for this type of dataset.

95% I got following the tutorial from here: https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/06_CIFAR-10.ipynb

That NN is defined with the prettytensor package as:

> with pt.defaults_scope(activation_fn=tf.nn.relu, phase=phase):
        y_pred, loss = x_pretty.\
            conv2d(kernel=5, depth=64, name='layer_conv1', batch_normalize=True).\
            max_pool(kernel=2, stride=2).\
            conv2d(kernel=5, depth=64, name='layer_conv2').\
            max_pool(kernel=2, stride=2).\
            flatten().\
            fully_connected(size=256, name='layer_fc1').\
            fully_connected(size=128, name='layer_fc2').\
            softmax_classifier(num_classes=num_classes, labels=y_true)



If I try to build a NN similar to the tensorflow one, I get up to 81% training / 79% test accuracy, then on the next (26th) iteration I get the accuracy drops to 25% training and 23% test, and stays there for the remainder of the training.

My attempt of building the tensorflow NN:

`const serial_size_t n_fmaps = 64;
    const serial_size_t n_fc1 = 256;
    const serial_size_t n_fc2 = 128;

    nn << convolutional_layer<relu>(32, 32, 5, 1, n_fmaps, padding::same)
        << max_pooling_layer<relu>(32, 32, n_fmaps, 2)

        << convolutional_layer<relu>(16, 16, 5, n_fmaps, n_fmaps, padding::same)
        << max_pooling_layer<relu>(16, 16, n_fmaps, 2)

        << fully_connected_layer<relu>(8 * 8 * n_fmaps, n_fc1)
        << fully_connected_layer<relu>(n_fc1, n_fc2)
        << fully_connected_layer<softmax>(n_fc2, 4);`

Apart from the data scaling issue, I can also see that the TF NN has 'batch normalization' layers after max_pool. How do I add those layers? I tried using batch_normalization_layer, but the compilation fails.

**Comment by swigicat** [comment]
It seems to be related to the batch normalization. I've added batch_normalization_layers after the convolutional ones and now I'm getting 97% training and 92% test accuracy on the 22nd epoch, but it falls to 25% again on the 23rd epoch. Any ideas on why this happens?

I tried playing with the batch, learning rate and number of epochs, but anything else other than the defaults - 10, 0.01 and 30 - either fails from the start or somewhere in the middle...

**Comment by edgarriba** [comment]
even reducing the learning rate?

**Comment by edgarriba** [comment]
what's your cost function in this case?

**Comment by swigicat** [comment]
Hi Edgarriba,
I'm using the default cross_entropy cost function that was in the cifar-10 example. I'm not sure what the tensorflow example is using though, couldn't find a reference anywhere.
With a 0.001 learning rate the accuracy after 30 epochs is 72% training and 67% test. If I let it run more, on epoch 105 I get to a maximum of 85% test accuracy, then it drops back to 25% on the next epoch. I will let it run tonight with 1e-4 learning rate and see what it does, since this is the 'learning rate' that TF uses, for 10k iterations.

**Comment by kecsap** [comment]
Interestingly, I experience the opposite with my own 1D datasets. tiny-dnn performs extremely well. Is there some kind of advanced mathematics why is it so awesome?

I put a bunch of effort to test Caffe and dynet these weeks and compare the performance with tiny-dnn. To my greatest surprise:
- tiny-dnn is very robust. It converges very well with the default learning parameters (adagrad). When I feed different kind of data, it provides at least the accuracy of the traditional classifiers (SVM, random forest), but usually they are outperformed by tiny-dnn.
- Caffe & dynet: Using _exactly_ the same gradient function and the network topology, they perform poorly. I mean really poorly. Instead of a decreasing loss over iterations/epoch, the loss function jumps back and forth, but it does not decrease over time. I tried different gradient function, learning rates, but the same crap. I also tried the Intel's fork of Caffe. These frameworks just do not work with 1D data. I suppose they work with 2D data because they are popular...

...or tiny-dnn is a hidden gem what the masses did not realize yet. :)

**Comment by swigicat** [comment]
No luck with learning rate = 1e-4 either, unfortunately :( Again it goes up to ~85% then drops to 25% around epoch 1000. Any idea what can cause this behavior? The learning rate is pretty small...

@kecsap it your data uniformly distributed along [0..1]? My data contains a lot of zeroes, and I'm not sure if this is an issue.

**Comment by kecsap** [comment]
Hi swigicat,

I am curious about your problem, but I would need to reproduce it. Can you upload and share the exact C++ source file and images what you have?

**Comment by kecsap** [comment]
I have some suggestions:

- Try leaky relu activation function instead of relu for fully connected layers. The networks converges a bit faster for me with this function.
- Try tan_h activation function for convolutional_layer. It will not boost the final accuracy, but the network converges faster in the beginning.
- Try identity instead of softmax in the last layer.

**Comment by swigicat** [comment]
Thanks for your suggestions kecsap, they improved the stability a bit, and I get 93% accuracy after 5 iterations with lr=0.01 on a new (bigger) dataset; on the 6th iteration it breaks again - but these results are adequate for the demo I'm building.

I tried changing a lot of stuff to get it to a point that I can leave it running over night with better parameters, but it just doesn't seem to work, at some point or other the learning breaks, and I'm not knowledgeable enough in this field to properly understand why it does that. I think I'm going to have to go with another library in the future, at least for my dataset this is too unstable :(

Thanks again for your help!

**Comment by kecsap** [comment]
Sorry to hear, we cannot help without providing example and dataset.

I never experienced collapse with tiny-dnn however you must track the train-test errors continuously. At least when the testing error starts to decrease, you should stop the training (called early stopping) because the training error will decrease, but the testing error will increase because of the overfitting. You can also try to increase the neuron count in the fully connected layers. The bigger a dataset the bigger network capacity is needed to learn it.
